Address #703: make syntax/data model error fallback clear (#704)
unicode-org/message-format-wg: spec/formatting.md

## Introduction

This document defines the behavior of a MessageFormat 2.0 implementation
when formatting a message for display in a user interface, or for some later processing.

To start, parse the _message_ from its syntax or create it from a data model description.

> If _Syntax Errors_ are encountered, emit those errors.

Reviewer: It seems like the deletion of "data model errors" makes it implementation-dependent whether data model errors are recoverable, or whether they should be treated exactly like syntax errors.

That's fine, but I would support opening a separate issue to discuss whether that's what we want, or whether the spec should require data model errors to be non-recoverable, as there seem to be different opinions.

> If the message being formatted has any _Syntax Errors_ or _Data Model Errors_,

Reviewer 2: Should keep the bit about emitting the error. I'm also not really sure about the "message has errors" language here; I think the previous language on this was less opinionated about error handling.

```suggestion
If the message being formatted has any _Syntax Errors_ or _Data Model Errors_,
an appropriate error MUST be emitted and
```

Author: I'm not sure what else we would really say here? 

We could also do:

```suggestion
If the _message_ being formatted contains a _Syntax Error_ or _Data Model Error_
emit an appropriate error.
The result of formatting a _message_ containing such an error is a single _fallback value_
that uses the _message_'s fallback string defined in the _formatting context_
or, if this is not available or empty, the U+FFFD REPLACEMENT CHARACTER `�`.
```

The result of formatting a _message_ with a _Syntax Error_ is a single _fallback value_
that uses the _message_'s fallback string defined in the _formatting context_

> or, if this is not available or empty, the U+FFFD REPLACEMENT CHARACTER `�`.

Reviewer 3: Nitpick:

```suggestion
or, if this is not available or empty, the `U+FFFD` `REPLACEMENT CHARACTER` `�`.
```

Reviewer 2: The suggestion doesn't match the style we otherwise use for character names in the spec.

## Introduction

This document defines the behavior of a MessageFormat 2.0 implementation
when formatting a message for display in a user interface, or for some later processing.

To start, parse the _message_ from its syntax or create it from a data model description.

If the message being formatted has any _Syntax Errors_ or _Data Model Errors_,

> an appropriate error MUST be emitted and

Reviewer: This seems like it's requiring eager evaluation. It's reasonable for a lazy implementation to have to report syntax errors even in expressions that are never used. But when it comes to data model errors, I'm not sure. For example, duplicate option name errors: do we want to require an implementation to error out on:

```
.local $foo = {horse :func one=1 two=2 one=1}
{{I don't use any variables!}}
```

even though `$foo` is never referenced? Maybe this is a weird message and should always be an error. I don't have a strong opinion, I just want to make sure it's considered.

Reviewer 4: One of the things I did in my mock is to discard variables that were never used. And since they were never used, I never even tried to check them further. I don't think we need to force errors to be omitted in those cases, although it could be useful to have a 'strict' mode that emitted warnings, etc.

BTW — and this is not _in any way_ suggesting a change in v45 — I have a somewhat different breakdown of errors, based information available at a given point. Categories 1-4 are all errors that can be handled through static analysis, before any runtime use.

1. Errors that can be detected syntactically, without reference to any function registry. 
  * That includes any BNF breakages
  * But other failures that don't require building a model or having any access to information about functions and their options:  a mismatch between number of keys in a variant and the number of selectors in the .match, no fallback variant, duplicate declarations or option names, ...
2. Errors that can be detected with reference to reference to the registry, but without having any input parameters
  * This includes illegal options, illegal option values, some illegal literals (that the function can't handle), .match with functions that are not selector functions, placeholders with functions that aren't formatting functions, etc.
3. Errors that require calling functions, but need no access to input parameters
  * Literals that can't be handled, but regex in the registry is not powerful enough to limit
  * Mismatches like .input {$person :personName ..} .local $date {$person :datetime ...), if the registry is not detailed enough to handle.
4. Errors that can be detected at call sites (in strongly-typed programming languages, depending on APIs)
   *  There are not enough input parameters, or the types are not accepted by the functions that will be called on them.
5. Runtime errors, that cannot be assessed except with particular input parameters
  * A function can't handle a particular input datatype, etc.

Author: @catamorphism  notes:

> This seems like it's requiring eager evaluation.

We have a very strong statement that we never require eager evaluation. As you note, syntax errors are not what is in question here. They are always errors and they have nothing to do with expression evaluation.

Of the data model errors, most are not related to eager vs. lazy evaluation--except duplicate option name and maybe missing selector annotation. That seems like something that should not be included in the MUST I suggest above. The others are structural errors with the message that have to be evaluated in order for pattern selection to run. This is in part why the text here was originally in the pattern selection section. 

Probably we should split syntax from data model error and push the DM errors back into pattern selection where they started. (Expression evaluation already has the one for duplicate option names)

@macchiati 

You're welcome to discard unused variables--including never evaluating the expressions assigned to them--in your implementation (we say this explicitly). However, of the data model errors (which are not related to anything in the function registry), as noted, only the duplication of options is really "looking inside" the expression. The missing selector annotation has to be checked because it is on the `.match` statement. Any message with a `.match` has to evaluate the selectors.

I have more thoughts, but not time tonight to delve into them.

> the result of formatting the _message_ MUST be a pattern resolving to a single _fallback value_

Reviewer: This implies that no syntax errors are recoverable, right? For example, that:

```
.local $bar {|foo|} {{{$bar}}}
```

formats to `�` and not `{$bar}`.

It would be an improvement to disambiguate the spec in that way, but I'm not sure if that's the behavior everyone wants?

Reviewer 2: I'm having a hard time figuring out a situation where it would be overall beneficial to allow for messages with syntax errors to format to anything other than `{�}` or something like `{msg-id}`, in particular if that behaviour is different between implementations.

What is the process generating such a broken message that we want to support?

> because no matching _variant_ is available.

The number of _keys_ in each _variant_ MUST equal the number of _selectors_.

> If it does not, emit a _Variant Key Mismatch_ error

Reviewer: This text also seems to written to allow the implementation to choose whether to recover from data model errors (an implementation that checks for them statically would never reach this point and would just use the fallback value for the result of formatting, and then this would be an assertion rather than a dynamic error).

Which is also fine, but reinforces my suggestion to open a separate issue to discuss that.

Reviewer: I created #735 so that, if this PR lands, we don't forget to discuss data model errors.
